File: solutions/typescript/minesweeper/1/minesweeper.ts
export function annotate(field: Array<string>): Array<string> {
  let result: Array<string> = []
  for (let i = 0; i < field.length; i++) {
    result.push('')
    for (let j = 0; j < field[i].length; j++) {
      let numBomb = 0;
      if (field[i][j] !== '*') {
        for (let h = -1; h <= 1; h++) {
          for (let w = -1; w <= 1; w++) {
            if ((i + h >= 0) && (i + h < field.length) && (j + w >= 0) && (j + w < field[i].length)) {
              if (field[i + h][j + w] === '*') {
                numBomb++
              }
            }
          }
        }
        if (numBomb > 0) {
          result[i] += numBomb.toString()
          numBomb = 0
        } else {
          result[i] += ' '
        }
      } else {
        result[i] += '*'
      }
    }
  }
  return result;
}

File: solutions/c/minesweeper/2/minesweeper.c
#include "minesweeper.h"

#include <stdlib.h>
#include <string.h>


static size_t last_rows = 0;

char **annotate(const char **minefield, const size_t rows)
{
    if (minefield == NULL)
        return NULL;

    last_rows = rows;

    size_t row_length = strlen(minefield[0]);
    char** result = calloc(rows, sizeof(char *));
    for(size_t i = 0; i < rows; i++)
    {
        result[i] = calloc(row_length + 1, sizeof(char));
        for(size_t j = 0; j < row_length; j++) {
            result[i][j] = '0';
        }
        result[i][row_length] = '\0';
    }

    for (size_t i = 0; i < rows; i++) {
        for(size_t j = 0; j < row_length; j++) {
            if (minefield[i][j] == '*') {
                result[i][j] = '*';
                // One Row up
                if (i > 0) {
                    if (j > 0 && minefield[i - 1][j - 1] != '*')
                        result[i - 1][j - 1]++;
                    //Center
                    if (minefield[i - 1][j] != '*')
                        result[i - 1][j]++;
                    // Right
                    if (j + 1 < row_length && minefield[i - 1][j + 1] != '*')
                        result[i - 1][j + 1]++;
                }
                // Left
                if (j > 0 && minefield[i][j - 1] != '*')
                    result[i][j - 1]++;
                // Right
                if (j + 1 < row_length && minefield[i][j + 1] != '*')
                    result[i][j + 1]++;

                // One Row Down

                if (i + 1 < rows) {
                    if (j > 0 && minefield[i + 1][j - 1] != '*')
                        result[i + 1][j - 1]++;

                    if (minefield[i + 1][j] != '*')
                        result[i + 1][j]++;

                    if (j + 1 < row_length && minefield[i + 1][j + 1] != '*')
                        result[i + 1][j + 1]++;
                }
            }
        }
    }

    // Replace 0 mines with whitespace

    for(size_t i = 0; i < rows; i++)
        for(size_t j = 0; j < row_length; j++)
            if (result[i][j] == '0')
                result[i][j] = ' ';

    return result;
}
